src/Compilers/Core/Portable/InternalUtilities/StackGuard.cs

```diff
@@ -29,5 +30,27 @@ public static void EnsureSufficientExecutionStack(int recursionDepth)
                 RuntimeHelpers.EnsureSufficientExecutionStack();
             }
         }
+
+        internal static bool TryEnsureSufficientExecutionStack(int recursionDepth, bool throwOnFailure)
+        {
+            try
+            {
+                EnsureSufficientExecutionStack(recursionDepth);
+                return true;
+            }
+            catch (InsufficientExecutionStackException) when (!throwOnFailure)
+            {
+                return false;
+            }
+        }
+
+        internal static TResult ExecuteOnNewExecutionStack<TResult>(Func<TResult> execute)
+        {
+            var task = Task.Run(() => execute());
+            // Wait on the task without inlining the task on this thread.
+            Task.WhenAny(task).Wait();
+            // Return result, propagating any exception.
```

Comment:
>propagating any exception [](start = 30, length = 25)

Are you sure that this doesn't result in the propagated exception being wrapped?

Reply:
`.GetAwaiter().GetResult()` is what's used by `await` and will not wrap the exception.  That's what the comment is referring to.

The line above it using `.Wait()` would wrap an exception if there was one, but the wait is on the `Task<Task>` returned from `WhenAny`; there won't be any exception there, even if `task` completes as faulted.